### PR TITLE
Add a new command function in Cypher named 'keys' that returns the property names

### PR DESCRIPTION
**Pull request by antonio-castellon**
...names from the properties of Node(s).

usage: *keys(n), where n=> node(s) result(s)*
example 1 : `MATCH (n) RETURN keys(n);`
example 2 : `MATCH (n) WITH keys(n) AS p UNWIND p AS x WITH DISTINCT x RETURN collect(x) AS SET;`

### CODE REVIEW

File: community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunction.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+package org.neo4j.cypher.internal.compiler.v2_2.commands.expressions
+
+import org.neo4j.cypher.internal.compiler.v2_2._
+import org.neo4j.cypher.internal.compiler.v2_2.executionplan.Effects
+import pipes.QueryState
+import symbols._
+import org.neo4j.cypher.internal.compiler.v2_2.spi.QueryContext
+import org.neo4j.graphdb.Node
+
+case class PropertiesFunction(nodeExpr: Expression) extends Expression {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunction.scala]: You might want to extend `NullInNullOutExpression` here.

File: community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunctionTest.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.mockito.Mockito._
+import org.mockito.invocation.InvocationOnMock
+import org.mockito.stubbing.Answer
+import org.neo4j.cypher.internal.commons.CypherFunSuite
+import org.neo4j.cypher.internal.compiler.v2_2._
+import org.neo4j.cypher.internal.compiler.v2_2.pipes.QueryStateHelper
+import org.neo4j.cypher.internal.compiler.v2_2.spi.QueryContext
+import org.neo4j.graphdb.Node
+
+class PropertiesFunctionTest extends CypherFunSuite {
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunctionTest.scala]: It would be nice to have a test for the case where there are no properties on the node.

Comment (antonio-castellon) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunctionTest.scala]: OK, what is the best option when I finish these updates ... 1) remove the pull and then 2) make a new query for a unique pull with an unique commit?

Comment (chrisvest) [inline community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/PropertiesFunctionTest.scala]: Just amend the commits in this pull-request.

**Comment by antonio-castellon** [comment]
Mmm. it´s my first time using Git in a community project and of course, as a commiter in `Neo4j` and `TeamCity Build` system... Someone can help me to clarify what is the reason about the message from TeamCity Build? because it's working on my local computer without any problem. :(

**Comment by thobe** [comment]
@antonio-castellon the failure was because you had not been added to the list of accepted contributors. This was since your email accepting our CLA got stuck in a spam filter. I have sorted that out and restarted the build.

**Comment by nawroth** [comment]
I should be added to the manual as well.
See https://github.com/neo4j/neo4j/blob/master/community/cypher/docs/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/FunctionsTest.scala
and https://github.com/neo4j/neo4j/blob/master/community/cypher/docs/cypher-docs/src/docs/dev/ql/functions/index.asciidoc

**Comment by davidegrohmann** [comment]
An acceptance test is missing, perhaps one can be added to `FunctionsAcceptanceTest`.

**Comment by antonio-castellon** [comment]
Dear @nawroth & Cia.

1) `FunctionsTest.scala` was done with a new test on my fork.

2) About the `index.asccidoc` ... I have no idea where I need to include my file `properties.asciidoc` into the folder structure.. I try to find other files in the same section `Collection Functions` without success. 

3) `PropertiesAcceptanceTest.scala` was done as @davidegrohmann demand.

Please, Could you guide me a little more in these (2) task? 
Thanks in advance
Antonio

**Comment by nawroth** [comment]
@antonio-castellon Those files are generated by the tests and end up in
_neo4j/community/cypher/docs/cypher-docs/target/docs/dev/ql/functions_.
Before the include happens, _src/docs_ and _target/docs_ are merged, that's why the simple include statements in the `index.asciidoc` document work.
Add it right after the include for the `labels` function for now:
https://github.com/neo4j/neo4j/blame/master/community/cypher/docs/cypher-docs/src/docs/dev/ql/functions/index.asciidoc#L88

**Comment by freethejazz** [comment]
I think `properties` is a bit ambiguous for what the function is actually doing (getting the names of named properties).  In the original issue, #1695, we discussed using `keys` as the function name, which I think is more precise and intuitive to those new to cypher.

What to people think? Is there any real danger of ambiguity?

**Comment by jexp** [comment]
keys() is the better name, please adapt

**Comment by antonio-castellon** [comment]
Yes, is true ... is much better to use keys(), I'm working on it.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+              assert(n.getPropertyKeys() === expectedProperties)
+
+            case _ =>
+              throw new AssertionError("assertThat used with result that is not a node")
+          }
+        }
+      }
+    }
+
+    insertNewCleanDatabase()
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala]: Why not having 2 different tests instead of cleaning the database here?

Comment (antonio-castellon) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala]: Ok, I take note.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.scalatest.Assertions
+import org.neo4j.graphdb.Node
+import org.scalautils.LegacyTripleEquals
+
+class PropertiesAcceptanceTest extends ExecutionEngineFunSuite
+with QueryStatisticsTestSupport with Assertions with CollectionSupport with LegacyTripleEquals {
+
+  test("Using_properties_in_RETURN_clauses") {
+    createNode()
+    assertThat("match (n) where id(n) = 0 RETURN properties(n)", List())
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala]: You can simplify the queries by removing the `where` clause, since the test db contains only one node there is no need to filter.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    createNode()
+    assertThat("match (n) where id(n) = 0 SET n.FOO='potato' RETURN properties(n)", List("FOO"))
+  }
+
+  private def assertThat(q: String, expectedProperties: List[String]) {
+    val result = execute(q).toList
+
+    graph.inTx {
+      if (result.isEmpty) {
+        val n = graph.getNodeById(0)
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/PropertiesAcceptanceTest.scala]: Same here, I think it is enough to fetch all the nodes instead having an assumption on the node id.

**Comment by freethejazz** [comment]
@antonio-castellon Can I help you out with any of this? I'm game for writing up some documentation and working on the minor changes to the neo4j browser that will be needed. What do you think?

**Comment by antonio-castellon** [comment]
All help will be welcome @freethejazz because I have some problems to compile the whole test code on Windows (https://github.com/neo4j/neo4j/issues/3883) and also in Linux (https://github.com/neo4j/neo4j/issues/3911). I prefer that someone help me to clarify what are the causes from this errors and allows me to compile all the code with the test cases... but if nobody can help me on it, then if you can add the `asciidoc` as @nawroth told me, I will happy to receive your help and close this part of the implementation.

**Comment by antonio-castellon** [comment]
Dear Neo4j Developes, 

Now I'm trying to extend this feature to include also the `Relationships` (not only `Nodes`) to be used with this function. I finish this new feature on my local computer but I have questions before to start to do all the Test Cases, because I have no idea how to affect to the rest of the code and I want to have answers from the code's owners.

```
case class KeysFunction(nodeOrRelationshipExpr: Expression) extends NullInNullOutExpression(nodeOrRelationshipExpr) {
[...]
 protected def calculateType(symbols: SymbolTable) = {
    nodeOrRelationshipExpr.evaluateType(CTNode, symbols)
    CTCollection(CTString)
}
```
1) What is the meaning of this function (I supposed that is only to filter the expressions according with the expected type of input data)..but I'm not sure about it.
2) I tested removing the `nodeOrRelationshipExpr.evaluateType(CTNode, symbols)` sentence and everything is working well... but, maybe we can do it more sparty to including both types (nodes and relationships) right? 

Also,
```
  override def localEffects = Effects.READS_NODES
```

1) What is the meaning of this function? It's used only when the object create a new data? have sense to include other effects in this case?

Thanks in advance and Have a nice day
Antonio

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  test("Using_keys_function_with_Not_Empty_result") {
+    createNode()
+    assertThat("match (n) SET n.FOO='potato' RETURN keys(n)", List("FOO"))
+  }
+
+  private def assertThat(q: String, expectedProperties: List[String]) {
+    val result = execute(q).toList
+
+    graph.inTx {
+      if (result.isEmpty) {
```
Comment (systay) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala]: I can't see any uses of this method with an empty result. Maybe simplify this assertion to do less? `getNodeById(0)` below looks very fragile.

**Comment by systay** [comment]
Hi @antonio-castellon,

Thanks so much for the work you have done! It makes me happy to see.

Let me try to answer your questions:

1) & 2) This is one of the remnants of the old legacy type system. These checks are all done by the ast-tree as part of semantic checking (e.g. https://github.com/neo4j/neo4j/blob/master/community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/ast/Property.scala#L31 ). Should be cleaned out any day now. You don't have to evaluate the type of the node - simply returning `CTCollection(CTString)` should be fine.

3) That is used to calculate when we need eagerness injected in the query plan. See here: https://github.com/neo4j/neo4j/blob/master/community/cypher/cypher-compiler-2.2/src/test/scala/org/neo4j/cypher/internal/compiler/v2_2/executionplan/AddEagernessIfNecessaryTest.scala. For the actions the `keys()` functions is doing now, `Effects.READS_NODES` seems to be exactly right.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala (hunk 6 above)
Comment (antonio-castellon) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala]: Me too, but @davidegrohmann told me to do a test with a Node without any property. Maybe I misunderstand him. 

About to the `getNodeById(0)`, you are right.. in the begining this code had ` where id(n) = 0 ` to catch exactly the code, but later was deleted. Now I'm preparing a new version of this feature with a best Test Cases.

Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala]: @systay You should look closer at the first test `"Using_keys_function_with_Empty_result"`, I am quite sure you get an empty result in that case.

+1 on getting rid off `getNodeById(0)`.

**Comment by antonio-castellon** [comment]
Hi @systay thanks all dev. group for this fantastic solution and for this excuse to learnt Scala with a real project. I hope that I will be able to collaborate more on it. ;)

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher
+
+import internal.helpers.CollectionSupport
+import org.scalatest.Assertions
+import org.neo4j.graphdb.{Relationship, Node}
+import org.scalautils.LegacyTripleEquals
+
+class KeysAcceptanceTest extends ExecutionEngineFunSuite  with QueryStatisticsTestSupport {
```
Comment (systay) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/KeysAcceptanceTest.scala]: It would be lovely, for the sake of completeness, to add a couple of tests where the node/rel is null (using `OPTIONAL MATCH`), and maybe a couple of expected failure situations and how the function is expected to handle them?

**Comment by systay** [comment]
It's looking really good now, @antonio-castellon. One more hygienic things - do you mind cleaning up the commit history for this feature? Happy to lend a hand if you need it.

**Comment by antonio-castellon** [comment]
well @systay I'm trying to find a solution to merge everything in one unique commit on the pull, but it's a little nightmare... I don't want to delete all my fork and start again from zero (with a local copy from the sopurces, of course). Exists any list of commands that I can do to, for example.. restore my fork to the origin/master point and then commit and push all the changes in one shot?

I find several methods to do it, but I have no idea what are the best (clean) solution. Could you help me on it, please?

**Comment by antonio-castellon** [comment]
What about to close this `pull` request and re-pull it again from my last fork version? it's an stupid way (nothing changes  : All the different commits appears again into the pull) or with this option only shows the last differences as a unique commit? Sorry, It's my first time using Git.

**Comment by systay** [comment]
I'll create that commit for you and share the URL when it's ready.

**Comment by antonio-castellon** [comment]
Thanks @systay , but also will be nice to learnt it for other future pulls ;) Could you send me your steps in a short way/description?

Thanks a lot !!

**Comment by systay** [comment]
Your branch was in a bit of a weird state, so I had to be insistent to get it to work. Normally using `git rebase -i` is not very difficult. Here is a good tutorial for this stuff: https://www.atlassian.com/git/tutorials/rewriting-history/git-commit--amend

**Comment by systay** [comment]
I pushed the merged commit to my repo here: https://github.com/systay/neo4j/tree/antonio

**Comment by antonio-castellon** [comment]
And.... what I need to do? maybe we can talk together directly on a chat or by email in order to not full this space with simple/stupid questions. Not? ;) [antonio dot  castellon at gmail dot com]

**Comment by systay** [comment]
I created a PR with the squashed commit here https://github.com/neo4j/neo4j/pull/3938. Hope that helps!

**Comment by boggle** [comment]
We are considering adding dynamic property lookup to Cypher; please comment on the proposal at

https://docs.google.com/document/d/1FPfGkgzhcRXVkleBLBsA92U94Mx4yafu3nO-Xf-NzsE/edit?usp=sharing
